### PR TITLE
Validations

### PR DESCRIPTION
**Pull request by jroodman**
This PR creates adds validations to the models for HackerNews. It also adds testing for these validations and all associations.

### CODE REVIEW

File: app/models/vote.rb [hunk 1]
```diff
@@ -2,6 +2,8 @@ class Vote < ApplicationRecord
 
   belongs_to :votable, polymorphic: true
 
+  validates :user_id, :votable_id, :votable_type, presence: true
+
```
Comment (rysospicy) [inline app/models/vote.rb]: I think you'll also want an application-level `uniqueness` validation here for the combination of `user_id`, `votable_id`, and `votable_type`.  Check out the bit on the `:scope` option in the [validation Rails docs](http://guides.rubyonrails.org/active_record_validations.html#uniqueness).

File: spec/models/vote_spec.rb [hunk 2]
```diff
@@ -1,6 +1,13 @@
 require 'rails_helper'
 
 RSpec.describe Vote, type: :model do
+
+  it { should belong_to(:votable) }
+
+  it { should validate_presence_of :user_id }
+  it { should validate_presence_of :votable_type }
+  it { should validate_presence_of :votable_id }
+
```
Comment (rysospicy) [inline spec/models/vote_spec.rb:28]: After adding the uniqueness validation to the model, be sure to cover it with a test here as well.

**Review by rysospicy — APPROVED** [review]
Looks good!  There's one more validation to address, but otherwise :100:

File: app/models/comment.rb [hunk 3]
```diff
@@ -1,11 +1,12 @@
 class Comment < ApplicationRecord
 
   belongs_to :user
+  belongs_to :parent, class_name: :Comment, foreign_key: :parent_comment_id
```
Comment (zporter) [inline app/models/comment.rb]: can a Comment also belong to a post / link?

File: spec/models/comment_spec.rb [hunk 4]
```diff
@@ -1,6 +1,16 @@
...
 RSpec.describe Comment, type: :model do
+
+  it { should belong_to(:user) }
+  it { should belong_to(:parent).class_name(:Comment).with_foreign_key(:parent_comment_id) }
+  it { should have_many(:children).class_name(:Comment).with_foreign_key(:parent_comment_id) }
+
+  it { should validate_presence_of :text }
+  it { should validate_presence_of :user_id }
+  it { should validate_presence_of :link_id }
+  it { should validate_presence_of :votes_count }
```
Comment (zporter) [inline spec/models/comment_spec.rb:13]: will `votes_count` eventually be a [counter cache](http://guides.rubyonrails.org/association_basics.html#options-for-belongs-to-counter-cache)?

**Review by zporter — CHANGES_REQUESTED** [review]
Left a couple of comments to address. Ping me in Slack when you've addressed them so I can re-review.  Great work!

File: spec/models/comment_spec.rb (hunk 4 above)
Comment (jroodman) [reply spec/models/comment_spec.rb:13]: yup!

File: app/models/comment.rb (hunk 3 above)
Comment (jroodman) [reply app/models/comment.rb]: Yup, good catch. Thanks!

File: spec/models/comment_spec.rb (hunk 4 above)
Comment (jroodman) [reply spec/models/comment_spec.rb:13]: Good catch on that as well, put off implementing that earlier and forgot to add it in.

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -1,6 +1,29 @@
...
+
+  it { should validate_length_of(:username).is_at_least(7).is_at_most(50) }
+  it { should validate_length_of(:password).is_at_least(8).is_at_most(256) }
+
+  it { should allow_value('email@test.com').for(:email) }
+  it { should allow_value('username12').for(:username) }
+  it { should_not allow_value('email').for(:email) }
+  it { should_not allow_value('email@test').for(:email) }
+  it { should_not allow_value('username').for(:username) }
+  it { should_not allow_value('username space').for(:username) }
```
Comment (zporter) [inline spec/models/user_spec.rb]: add a test for a username with 1 number

**Review by zporter — APPROVED** [review]
nice work! Add one more test around a username with 1 number and receive a 👍

**Pull request merged**